keystore/src/transaction.rs

```diff
@@ -0,0 +1,340 @@
+            Operation::Store(Entity::PersistedMlsGroup(_)) | Operation::Remove(EntityId::PersistedMlsGroup(_)) => {
+                PersistedMlsGroup::COLLECTION_NAME
+            }
+            Operation::Store(Entity::PersistedMlsPendingGroup(_))
+            | Operation::Remove(EntityId::PersistedMlsPendingGroup(_)) => PersistedMlsPendingGroup::COLLECTION_NAME,
+            Operation::Store(Entity::MlsPendingMessage(_)) | Operation::Remove(EntityId::MlsPendingMessage(_)) => {
+                MlsPendingMessage::COLLECTION_NAME
+            }
+            Operation::Store(Entity::E2eiEnrollment(_)) | Operation::Remove(EntityId::E2eiEnrollment(_)) => {
+                E2eiEnrollment::COLLECTION_NAME
+            }
+            Operation::Store(Entity::E2eiRefreshToken(_)) | Operation::Remove(EntityId::E2eiRefreshToken(_)) => {
+                E2eiRefreshToken::COLLECTION_NAME
+            }
+            Operation::Store(Entity::E2eiAcmeCA(_)) | Operation::Remove(EntityId::E2eiAcmeCA(_)) => {
+                E2eiAcmeCA::COLLECTION_NAME
+            }
+            Operation::Store(Entity::E2eiIntermediateCert(_))
+            | Operation::Remove(EntityId::E2eiIntermediateCert(_)) => E2eiIntermediateCert::COLLECTION_NAME,
+            Operation::Store(Entity::E2eiCrl(_)) | Operation::Remove(EntityId::E2eiCrl(_)) => E2eiCrl::COLLECTION_NAME,
+        }
+    }
+}
+
+/// This represents a transaction, where all operations will be done in memory and committed at the
+/// end
+#[derive(Clone)]
+pub struct KeystoreTransaction {
+    /// Reference to the connection
+    conn: Connection,
+    // ideally we'd have boxed dyn `EntityMlsExt` here, but this trait cannot be used as object
+    // unfortunatelly we have to live with a lot of boilerplate code because of that
+    operations: Arc<RwLock<VecDeque<Operation>>>,
+}
+
+impl KeystoreTransaction {
+    pub fn new(conn: Connection) -> Self {
+        Self {
+            conn,
+            operations: Default::default(),
+        }
+    }
+
+    /// Persists all the operations in the database. It will effectively open a transaction
+    /// internally, perform all the buffered operations and commit.
+    ///
+    /// TODO: currently only MLS is supported. Implement for proteus. For that, remove the default
+    /// save, insert and delete functions in the `EntityBase` trait
+    /// FIXME: implement a transaction wrapper for the wasm platform to await on the transaction
+    pub async fn commit(&self) -> Result<(), CryptoKeystoreError> {
+        // we don't necessarily need the write lock here, but since we don't want additional
+        // operations being added to the queue, we get an exclusive lock
+        let mut operations = self.operations.write().await;
+        let mut conn = self.conn.borrow_conn().await?;
+        cfg_if::cfg_if! {
+            if #[cfg(target_family = "wasm")] {
+                let tables = operations.iter().map(|op| op.scan_collection_name()).collect::<Vec<_>>();
```

Comment:
What happens if this vec contains the same string twice? Is this okay for idb?

Reply:
Fixed in: https://github.com/wireapp/core-crypto/pull/695/commits/9d7bae311457113a73dffd873dd9c78c8e9e23fc